gnc/sub8_perception/nodes/buoy_2d.py

```diff
@@ -21,61 +22,92 @@ class BuoyFinder:
-        self.last_t = None
-
-        self.observations = deque()
-        self.pose_pairs = deque()
+        self.ppf = None
+        self.multi_obs = None
+        self.max_observations = 50
+        self._id = 0  # Only for display
 
         self.rviz = rviz.RvizVisualizer()
 
-        self.pose2d_service = rospy.Service('vision/buoys/2D', VisionRequest2D, self.request_buoy)
-        self.pose_service = rospy.Service('vision/buoys/pose', VisionRequest, self.request_buoy3d)
-
-        self.image_sub = sub8_ros_tools.Image_Subscriber('/stereo/right/image_rect_color', self.image_cb)
-        self.image_pub = sub8_ros_tools.Image_Publisher('/vision/buoy_2d/target_info')
-
-        # Occasional status publisher
-        self.timer = rospy.Timer(rospy.Duration(1.0), self.publish_target_info)
-
         rospack = rospkg.RosPack()
         boost_path = os.path.join(
             rospack.get_path('sub8_perception'),
             'sub8_vision_tools',
+            'machine_learning',
             'classifiers',
-            'boost.cv2'
+            'red_gentle_20tree_9depth.dic'
         )
 
         self.boost = cv2.Boost()
         rospy.loginfo("Loading boost")
         self.boost.load(boost_path)
         rospy.loginfo("Boost loaded")
 
+        self.observations = {
+            'red':deque(), 
+            'yellow':deque(), 
+            'green':deque()
+        }
+        self.pose_pairs = {
+            'red':deque(), 
+            'yellow':deque(), 
+            'green':deque()
+        }
         self.buoys = {
             'green': '/color/buoy/green',
             'red': '/color/buoy/red',
             'yellow': '/color/buoy/yellow',
         }
+        self.last_t = {
+            'green': None,
+            'red': None,
+            'yellow': None
+        }
 
-        self.ppf = None
-        self.multi_obs = None
-
+        # For displaying each buoy in rviz
         self.draw_colors = {
             'green': (0.0, 1.0, 0.0, 1.0),
             'red': (1.0, 0.0, 0.0, 1.0),
             'yellow': (1.0, 1.0, 0.0, 1.0),
         }
+        self.visual_id = {
+            'green': 0,
+            'red': 1,
+            'yellow': 2,
+        }
+
+        self.image_sub = sub8_ros_tools.Image_Subscriber('/stereo/right/image_raw', self.image_cb)
+        self.image_pub = sub8_ros_tools.Image_Publisher('/vision/buoy_2d/target_info')
+        self.mask_pub = sub8_ros_tools.Image_Publisher('/vision/buoy_2d/mask')
+
+        # Occasional status publisher
+        self.timer = rospy.Timer(rospy.Duration(1), self.publish_target_info)
+
+        self.toggle = rospy.Service('vision/buoys/search', SetBool, self.toggle_search)
+        self.pose2d_service = rospy.Service('vision/buoys/2D', VisionRequest2D, self.request_buoy)
+        self.pose_service = rospy.Service('vision/buoys/pose', VisionRequest, self.request_buoy3d)
+
+    def toggle_search(self, srv):
+        if srv.data:
+            rospy.loginfo("MARKER - Looking for markers now.")
```

Comment:
Whoops, copied and pasted from the channel marker one. Forgot to change the name